src/Consumer.php

```diff
@@ -0,0 +1,161 @@
+                $this->getChannel()->wait(
+                    null,
+                    false,
+                    $this->getConsumerPropertyBasedOnInput($properties, 'timeout', 0)
+                );
+            }
+        } catch (AMQPTimeoutException) {
+            return true;
+        }
+
+        return true;
+    }
+
+    public function setCallback($callback): static
+    {
+        if (! is_callable($callback)) {
+            throw new InvalidConfigurationException("Callback $callback is not callable.");
+        }
+
+        $this->callback = $callback;
+
+        return $this;
+    }
+
+    /**
+     * @throws Throwable
+     */
+    public function handle(AMQPMessage $message): void
+    {
+        if (! is_callable($this->callback)) {
+            throw new InvalidConfigurationException("Callback $this->callback is not callable.");
+        }
+
+        try {
+            call_user_func($this->callback, $message, $this);
+            $this->acknowledge($message);
+        } catch (Throwable $e) {
+            $message->getChannel()->basic_reject(
+                $message->getDeliveryTag(),
+                true
+            );
+
+            throw $e;
+        }
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function setupQueues(): void
+    {
+        $queuesConfig = $this->getConsumerProperty('queues');
+        foreach ($queuesConfig as $queueConfig) {
+            $this->declareQueue($queueConfig['queue']);
+        }
+    }
+
+    private function acknowledge(AMQPMessage $message): void
+    {
+        $message->getChannel()->basic_ack($message->getDeliveryTag());
+
+        if ($message->body === 'quit') {
+            $message->getChannel()->basic_cancel($message->getConsumerTag());
+        }
+    }
+
+    private function declareQueue(string $queue): void
+    {
+        $queues = $this->getConsumerProperty('queues');
+
+        if (empty($queues)) {
+            throw new InvalidArgumentException("Queue '$queue' is not defined");
+        }
+
+        $queuesMap = Arr::keyBy($this->getConsumerProperty('queues'), 'queue');
+
+        if (! isset($queuesMap[$queue])) {
+            throw new InvalidArgumentException("Queue '$queue' is not defined");
+        }
+
+        $queueConfig = $queuesMap[$queue];
+        $this->getChannel()->queue_declare(
+            $queue,
+            $queueConfig['passive'] ?? false,
+            $queueConfig['durable'] ?? true,
+            $queueConfig['exclusive'] ?? false,
+            $queueConfig['auto_delete'] ?? false,
+            $queueConfig['nowait'] ?? false,
+            $queueConfig['properties'] ?? [],
+        );
+
+        foreach ($queueConfig['bindings'] as $bindingConfig) {
+            $this->getChannel()->queue_bind(
+                $queueConfig['queue'],
+                $bindingConfig['exchange'],
+                $bindingConfig['routingKey'] ?? '',
+            );
+        }
+    }
```

Comment:
**[blocker]** Isn’t the same queue being redeclared everytime a message is consumed? I imagine that might cause issues, but I might be mistaken.

I would validate this by creating a test case that calls `Consumer->consume()` several times and verifies that messages in the queue are still retrievable and no messages are getting lost.

Reply:
@DanBondarenko the queue declaration is called once when the process starts working. I tested the behavior and the declaration of the already existing queue will not end up with cleanup. 

In addition, if you declare the queue with the same name and parameters as the existing queue, RabbitMQ will simply return a confirmation message indicating that the queue already exists. This confirmation message can be useful for verifying that the queue has been properly configured and is ready to use.

Reply:
A-ha, I see. Sorry for the confusion and thank you for the clarification! :) 

Resolving